### PR TITLE
introduce internal packages for placeholders, path utility functions, globbing

### PR DESCRIPTION
**Pull request by alexanderwilling**
at the same time:
- rewrote globbing
- rewrote path splitting
- rewrote pattern resolving

new packages:
- placeholders
- paths
- stringz
- shared

integration tests pass: http://awsbuild.phraseapp.io/job/phraseapp-integration/710

### CODE REVIEW

File: internal/paths/glob.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+// Glob replaces placeholders with * and supports * and ** globbing according to docs.
+func Glob(pattern string) (matches []string, err error) {
+	pattern = filepath.Clean(pattern)
+
+	err = ValidatePattern(pattern)
+	if err != nil {
+		return nil, fmt.Errorf("invalid pattern '%s': %s", pattern, err)
+	}
+
+	if strings.Contains(pattern, "/**/") {
```
Comment (sacry-dyn) [inline internal/paths/glob.go]: (Question|Blocker) does this work with windows?

File: internal/placeholders/placeholders.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	"fmt"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/phrase/phraseapp-client/internal/stringz"
+)
+
+var (
+	AnyPlaceholderRegexp = regexp.MustCompile("<(locale_name|tag|locale_code)>")
```
Comment (sacry-dyn) [inline internal/placeholders/placeholders.go]: (Question) Why capitalized?

File: internal/placeholders/placeholders.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+	// drop first element, which is the entire string s wich match name ""
+	matches, matchNames = matches[1:], matchNames[1:]
+
+	values := map[string]string{}
+	for i, match := range matches {
+		placeholder := matchNames[i]
+		if value, ok := values[placeholder]; ok {
+			if match != value {
+				return nil, fmt.Errorf("string %q does not match pattern %q: placeholder %q is used twice with different values", s, patternRE, placeholder)
```
Comment (sacry-dyn) [inline internal/placeholders/placeholders.go]: (Question) How can this error come up? Is this something that the pre refactor logic did as well?

File: internal/paths/glob.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+		parts := strings.Split(pattern, "/**/")
+		basePattern, endPattern := parts[0], parts[1]
+
+		baseCandidates, err := filepath.Glob(basePattern)
+		if err != nil {
+			return nil, fmt.Errorf("invalid pattern '%s': %s", pattern, err)
+		}
+
+		for _, base := range directoriesOnly(baseCandidates) {
```
Comment (sacry-dyn) [inline internal/paths/glob.go]: 👍 The globing looks much better. If this is rolled out you need to be aware of any bugsnag issues coming up. We have a lot of tests but file handling cannot be fully tested for any system. The old algorithm did "work". Just be aware and heads up that you changed it.

File: internal/placeholders/placeholders.go (hunk 3 above)
Comment (alexanderwilling) [inline internal/placeholders/placeholders.go]: Example: you have a pattern with placeholders like `translations/<locale_code>/project_foo/<locale_code>.yml` and provide the path `translations/en/project_foo/de.yml`. What value should the `<locale_code>` placeholder be resolved to? The old code simply overwrote previously found values for a placeholder and the last one was used: https://github.com/phrase/phraseapp-client/blob/62c380c5f477ed7d204cdf3d88010a3ce7a16c88/push.go#L483

File: internal/placeholders/placeholders.go (hunk 2 above)
Comment (alexanderwilling) [inline internal/placeholders/placeholders.go]: fixed 👍

File: internal/paths/glob.go (hunk 1 above)
Comment (alexanderwilling) [inline internal/paths/glob.go]: good point, fixed 👍

File: internal/paths/glob.go (hunk 4 above)
Comment (alexanderwilling) [inline internal/paths/glob.go]: There are some tests for the Glob function, but I agree and will look out for real-world trouble. Pretty sure though that the old algorithm was even more bug-prone 🙃

**Comment by alexanderwilling** [comment]
new integration tests: http://awsbuild.phraseapp.io/job/phraseapp-integration/710

**Comment by sacry-dyn** [comment]
@alexanderwilling lgtm! Nice changes :)

**Pull request merged**